Fix method definition cop

kuahyeow opened the pull request:
Fixes https://github.com/rubocop/rubocop-rake/issues/42

Fixes `ClassDefinitionInTask`, and `MethodDefinitionInTask` to allow classes, modules and methods to defined inside tasks

tkuah committed:
Fix MethodDefinitionInTask to allow methods defined in task

As methods defined in task do not pollute top level

For example, this rakefile below will show that `do_something` does not
appear in top level methods

```
task :a do
  def do_something
    puts 'a'
  end

  do_something
end

task :b do
  def do_something
    puts 'b'
  end

  do_something
end

puts methods.include?(:do_something)
```

tkuah committed:
Fix ClassDefinitionInTask to allow class defined in task

As classes defined in task do not pollute top level.

For example, this Rakefile shows class C in task
does not pollute top level

```
task :a do
  class C
    def do_something
      puts 'a'
    end
  end

  C.new.do_something
end

task :b do
  class C
    def do_something
      puts 'b'
    end
  end

  C.new.do_something
end

puts Object.const_defined?('C')
```

tkuah committed:
Remove un-used helper method

kuahyeow commented:
@koic  Is the configured Circle CI still working. Am getting the error below:

```
 CircleCI Pipeline — Could not find a usable config.yml, you may have revoked the CircleCI OAuth app. 
```

pboling reviewed (commented):
This is correct.  The difference between which definition of a method or class wins is according to lexical order.  The only one that works properly, without polluting the global Ruby Object namespace is inside a task, (whether or not it is inside a namespace is irrelevant).

pboling commented on lib/rubocop/cop/rake/class_definition_in_task.rb:
```diff
@@ -3,13 +3,13 @@
+      # Detects class or module definition in a namespace,
       # because it is defined to the top level.
-      # It is confusing because the scope looks in the task or namespace,
+      # It is confusing because the scope looks in the namespace,
```
```suggestion
      # It is confusing because the scope appears as if it is in the namespace,
```

pboling replied:
@kuahyeow This grammatical issue still needs to be fixed.  It is incorrect to say it "looks in the namespace", which may imply to some that the namespace is relevant.  It makes more sense to say it "appears as if it is in the namespace"... when in fact it is not.

pboling commented on lib/rubocop/cop/rake/class_definition_in_task.rb:
```diff
@@ -33,7 +33,7 @@ class ClassDefinitionInTask < Base
         def on_class(node)
           return if Helper::ClassDefinition.in_class_definition?(node)
-          return unless Helper::TaskDefinition.in_task_or_namespace?(node)
+          return unless Helper::TaskDefinition.in_namespace?(node)
```
Actually this is the wrong test.  It should be "not in task".  If not in task then it is using the Global Ruby Object namespace.  It makes no difference if it is in a namespace or not.

pboling replied:
Testing if in a namespace here may have been allowing you to avoid a test for the Global Object Space, but that's really what you want.  According to this Cop's intention:

* Class, Module and Method definitions are OK _iff_ defined at the top-level OR inside a task.
* Class, Module, and Method definitions are not OK if defined inside a namespace.

I don't think it can accurately capture that intend without a test for the definitions at the top level of the file.

kuahyeow replied:
@pboling Sorry, I dropped this completely. Must have slipped through the cracks

> I don't think it can accurately capture that intent without a test for the definitions at the top level of the file.

There are such tests, which were pre-existing, please refer to:

- `does not register an offense to `def` at the top level`
- `allows class definition outside task`

Given this cop's intention is to only highlight "surprising" behavior, I will leave this logic as is.

Though I also see that the _name_ of this cop is in-correct. Not sure how this is normally handled (is it breaking?) but I will push a rename for now.

pboling commented on lib/rubocop/cop/rake/class_definition_in_task.rb:
```diff
@@ -33,7 +33,7 @@ class ClassDefinitionInTask < Base
         def on_class(node)
           return if Helper::ClassDefinition.in_class_definition?(node)
-          return unless Helper::TaskDefinition.in_task_or_namespace?(node)
+          return unless Helper::TaskDefinition.in_namespace?(node)
```
```suggestion
          return unless Helper::TaskDefinition.in_task?(node)
```

pboling commented on lib/rubocop/cop/rake/helper/task_definition.rb:
```diff
@@ -8,17 +8,17 @@ module TaskDefinition
+          def_node_matcher :namespace?, <<-PATTERN
             (block
-              (send _ {:task :namespace} ...)
+              (send _ {:namespace} ...)
```
```suggestion
              (send _ {:task} ...)
```

pboling commented on lib/rubocop/cop/rake/helper/task_definition.rb:
```diff
@@ -8,17 +8,17 @@ module TaskDefinition
           PATTERN
 
-          def in_task_or_namespace?(node)
+          def in_namespace?(node)
```
```suggestion
          def in_task?(node)
```

pboling commented on lib/rubocop/cop/rake/helper/task_definition.rb:
```diff
@@ -8,17 +8,17 @@ module TaskDefinition
           extend self
 
-          def_node_matcher :task_or_namespace?, <<-PATTERN
+          def_node_matcher :namespace?, <<-PATTERN
```
```suggestion
          def_node_matcher :task?, <<-PATTERN
```

pboling commented on lib/rubocop/cop/rake/helper/task_definition.rb:
```diff
@@ -8,17 +8,17 @@ module TaskDefinition
+          def in_namespace?(node)
             node.each_ancestor(:block).any? do |a|
-              task_or_namespace?(a)
+              namespace?(a)
```
```suggestion
              task?(a)
```

pboling commented on spec/rubocop/cop/rake/class_definition_in_task_spec.rb:
```diff
@@ -1,19 +1,37 @@
-      task :foo do
+      namespace 'foo' do
         module M
         ^^^^^^^^ Do not define a module in rake task, because it will be defined to the top level.
```
```suggestion
        ^^^^^^^^ Do not define a module in rake namespace, because it will be defined at the top level.
```

pboling commented:
Ping!  Just ran into this false positive again, and came looking to see if any activity here.

pboling reviewed (commented):
LGTM! ❤️

tkuah committed:
Rename class definition cop to reflect fixed behaviour

Since the cop is now fixed to warn against class in namespaces, update
the cop name accordingly.

tkuah committed:
Rename method definition cop to reflect fixed behaviour

Since the cop is now fixed to warn against method in namespaces, update
the cop name accordingly.

q3aiml commented:
Is this correctly handling when a task is nested inside a namespace? I believe that should be allowed, but the spec doesn't seem to cover it and it looks like it is disallowed.

For example this test fails:
```
  it 'does not register an offense to `def` in a task when wrapped in a namespace' do
    expect_no_offenses(<<~RUBY)
      namespace :bar do
        task :foo do
          def helper_method
            do_something
          end
        end
      end
    RUBY
  end
```

tkuah committed:
Allow definitions within a task that is within a namespace.

This is similar to definitions within a task (without a namespace).

kuahyeow commented:
That's true. I have now pushed a fix for this.

> Is this correctly handling when a task is nested inside a namespace? I believe that should be allowed, but the spec doesn't seem to cover it and it looks like it is disallowed.
> 
> For example this test fails:
> 
> ```
>   it 'does not register an offense to `def` in a task when wrapped in a namespace' do
>     expect_no_offenses(<<~RUBY)
>       namespace :bar do
>         task :foo do
>           def helper_method
>             do_something
>           end
>         end
>       end
>     RUBY
>   end
> ```

pboling reviewed (commented):
LGTM

pboling commented:
@kuahyeow The build passed before your last fix, but now it failing on Ruby >= 2.7.  Any ideas?

kuahyeow commented:
Looks like it's failing on unrelated rubocop errors....

```
rubocop
Inspecting 25 files
............C.......C....

Offenses:

lib/rubocop/cop/rake/helper/task_name.rb:11:30: C: [Correctable] InternalAffairs/NodeFirstOrLastArgument: Use #first_argument instead of #arguments[0].
            first_arg = node.arguments[0]
                             ^^^^^^^^^^^^
spec/rubocop/cop/rake/desc_spec.rb:4:6: C: [Correctable] InternalAffairs/ExampleDescription: Description does not match use of expect_offense.
  it 'register an offense for task on the top level' do
     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
spec/rubocop/cop/rake/desc_spec.rb:16:6: C: [Correctable] InternalAffairs/ExampleDescription: Description does not match use of expect_offense.
  it 'register an offense for task with block in a block' do
     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
spec/rubocop/cop/rake/desc_spec.rb:33:6: C: [Correctable] InternalAffairs/ExampleDescription: Description does not match use of expect_offense.
  it 'register an offense for task in kwbegin' do
     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

25 files inspected, 4 offenses detected, 4 offenses autocorrectable
rake aborted!
Command failed with status (1): [rubocop...]
/home/circleci/project/Rakefile:36:in `block in <top (required)>'
/home/circleci/.rubygems/gems/rake-12.3.3/exe/rake:27:in `<top (required)>'
/usr/local/bin/bundle:25:in `load'
/usr/local/bin/bundle:25:in `<main>'
Tasks: TOP => default => rubocop
(See full trace by running task with --trace)

Exited with code exit status 1
```

ydakuka commented:
@kuahyeow could you please rebase the branch? The CI specs should pass as expected now.

pboling commented:
I think I missed this update somehow...
```
  it 'does not register an offense to `def` in a task when wrapped in a namespace' do
    expect_no_offenses(<<~RUBY)
      namespace :bar do
        task :foo do
          def helper_method
            do_something
          end
        end
      end
    RUBY
  end
  ```

This is incorrect.
  
Methods can be defined inside a `task` block **only**.  If they are defined inside a `namespace` block they will pollute the global namespace.

The test should be changed to register an offense.

CC @q3aiml @kuahyeow

kuahyeow commented:
@ydakuka @pboling Rebased now ! The rspecs workflow now requires approval from a maintainer to run.

pboling commented:
@koic Please approve the workflow!

pboling commented on lib/rubocop/cop/rake/class_definition_in_namespace.rb:
```diff
@@ -3,13 +3,13 @@
+      # Detects class or module definition in a namespace,
       # because it is defined to the top level.
-      # It is confusing because the scope looks in the task or namespace,
+      # It is confusing because the scope looks in the namespace,
```
@kuahyeow This grammatical issue still needs to be fixed. It is incorrect to say it "the scope looks in the namespace", which may imply to some that the namespace is relevant to the scoping. It makes more sense to say it "appears as if it is in the namespace"... when in fact it is not.

```suggestion
      # It is confusing because the scope appears as if it is in the namespace,
```

pboling reviewed (commented):
Needs a minor grammatical fix

tkuah committed:
Clarify code comment

kuahyeow commented:
Thanks @pboling. I have updated the comment now. Please take a look !

On Sat, 5 Apr 2025 at 13:52, Peter Boling ***@***.***> wrote:

> ***@***.**** commented on this pull request.
>
> Needs a minor grammatical fix
> ------------------------------
>
> In lib/rubocop/cop/rake/class_definition_in_namespace.rb
> <https://github.com/rubocop/rubocop-rake/pull/44#discussion_r2029595732>:
>
> >        # because it is defined to the top level.
> -      # It is confusing because the scope looks in the task or namespace,
> +      # It is confusing because the scope looks in the namespace,
>
> @kuahyeow <https://github.com/kuahyeow> This grammatical issue still
> needs to be fixed. It is incorrect to say it "the scope looks in the
> namespace", which may imply to some that the namespace is relevant to the
> scoping. It makes more sense to say it "appears as if it is in the
> namespace"... when in fact it is not.
> ⬇️ Suggested change
>
> -      # It is confusing because the scope looks in the namespace,
> +      # It is confusing because the scope appears as if it is in the namespace,
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/rubocop/rubocop-rake/pull/44#pullrequestreview-2744386163>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAECU2FTZUXA4FCTBFQWFD2X4SLLAVCNFSM6AAAAABXWCUPDWVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDONBUGM4DMMJWGM>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

pboling reviewed (commented):
LGTM 💯

tylerhunt commented:
This one bit me today. Hoping for a merge and release soon.

rossme commented on lib/rubocop/cop/rake/method_definition_in_namespace.rb:
```diff
@@ -30,12 +30,14 @@ module Rake
       #   task :foo do
```
Thanks for this PR @kuahyeow 🙏 

I don't think "expected behavior" is necessarily the same as a "good" example. It only moves the method definition but still leaves it at the top level of the file, which introduces the same problem of polluting globally. We've already shown a "good" example by adding it in the `task` directly.

Not sure what you think, but I would still find this misleading particularly if I was only running a single rake file, and there was a duplicate method definition in a later rake file.

I would remove this example, and add a cop for the top level method definition. What do you think?

pboling replied:
I agree. Splitting the example so that only definitions inside the task are considered "good" would be good ;)

rossme replied:
I do see your point about the top level method and keeping this PR lean 👍 

I would still (gently) argue for adding a cop for any method outside of the `task` block seeing as we are only checking specific files ending in `.rake` and the specific quirks and features of Rake files.

Also removing the `good` example in a later PR if necessary also makes sense.

corsonknowles replied:
Changing it from `# good` to `# okay` is the Rubocop doc way to distinguish what the cop accepts vs best practices

rossme commented on config/default.yml:
```diff
@@ -4,8 +4,8 @@ Rake:
-Rake/ClassDefinitionInTask:
-  Description: 'Do not define a class or module in rake task, because it will be defined to the top level.'
+Rake/ClassDefinitionInNamespace:
+  Description: 'Do not define a class or module in a rake namespace, because it will be defined to the top level.'
```
```suggestion
  Description: 'Do not define a class or module outside of a rake task, because it will be defined to the top level.'
```

kuahyeow commented on config/default.yml:
```diff
@@ -24,7 +24,7 @@ Rake/DuplicateTask:
-Rake/MethodDefinitionInTask:
-  Description: 'Do not define a method in rake task, because it will be defined to the top level.'
+Rake/MethodDefinitionInNamespace:
+  Description: 'Do not define a method in a rake namespace, because it will be defined to the top level.'
```
Since we are not changing the logic (see https://github.com/rubocop/rubocop-rake/pull/44#discussion_r2281043346)  - I will not update this message
